File: java_solutions/42.java
class Solution {
    public int trap(int[] height) {
        int [] LH_MAX = new int[height.length]  ;
        int seen_max_height = -1 ; 
        for(int i = 0 ; i < height.length ; i++){
            seen_max_height = Math.max(seen_max_height , height[i]) ; 
            LH_MAX[i] = seen_max_height ; 
        }
        int water_trapped = 0 ; 
        seen_max_height = -1 ; 
        for(int i = height.length-1 ; i >= 0 ; i--){
            seen_max_height = Math.max(seen_max_height , height[i]) ; 
            // System.out.println("H: "+ height[i] + " L: "+ LH_MAX[i] + " R: " + seen_max_height);
            water_trapped += Math.abs(height[i] - Math.min(seen_max_height , LH_MAX[i]));
        }
        return water_trapped ; 
    }
}

File: rust_solutions/42.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        use std::cmp::{max,min} ; 
        let mut LH_MAX = Vec::new() ; 
        let mut seen_max_height : i32 = -1 ; 
        for h in height.iter() {
            seen_max_height = max(seen_max_height , *h) ; 
            LH_MAX.push(seen_max_height) ; 
        }
        seen_max_height = -1 ; 
        let mut water_trapped : i32 = 0 ;
        for (idx,h) in height.iter().enumerate().rev() {
            seen_max_height = max(seen_max_height , *h) ; 
            water_trapped += (*h - min(LH_MAX[idx],seen_max_height)).abs() ;
        }
        water_trapped
    }
}
